File: LeetCode/Search/Rust/Leetcode875-koko-eating-bananas.rs
impl Solution {
    pub fn min_eating_speed(piles: Vec<i32>, h: i32) -> i32 {
        let (mut l,mut r) = (1,-1);
        for &pile in &piles {
            r = r.max(pile);
        }
        let check = |k:i32| -> bool {
            let mut cnt = 0;
            for pile in &piles {
                cnt += (pile + k - 1) / k;
            }
            cnt <= h
        };
        while l < r {
            let mid = (l + r) >> 1;
            if check(mid) {
                r = mid;
            }else{
                l = mid + 1
            }
        }
        l
    }
}

File: LeetCode/Search/python/Leetcode875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)

        def check(k):
            cnt = 0
            for pile in piles:
                if pile <= k: cnt += 1
                else: cnt += (pile + k - 1) // k
            return cnt <= h

        while l < r:
            mid = (l + r) >> 1
            if check(mid): r = mid
            else: l = mid + 1
        return l
